### PR TITLE
Validate a joining node's version with version of existing cluster nodes

### PR DESCRIPTION
**Pull request by bleskes**
When a node tries to join a cluster, it goes through a validation step to make sure the node is compatible with the cluster. Currently we validation that the node can read the cluster state and that it is compatible with the indexes of the cluster. This PR adds validation that the joining node's version is compatible with the versions of existing nodes. Concretely we check that:

1) The node's min compatible version is higher or equal to any node in the cluster (this prevents a too-new node from joining)
2) The node's version is higher or equal to the min compat version of all cluster nodes (this prevents a too old join where, for example, the master is on 5.6, there's another 6.0 node in the cluster and a 5.4 node tries to join).
3) The node's major version is at least as higher as the lowest node in the cluster. This is important as we use the minimum version in the cluster to stop executing bwc code for operations that require multiple nodes. If the nodes are already operating in "new cluster mode", we should prevent nodes from the previous major to join (even if they are wire level compatible). This does mean that if you have a very unlucky partition during the upgrade which partitions all old nodes which are also a minority / data nodes only, the may not be able to re-join the cluster. We feel this edge case risk is well worth the simplification it brings to BWC layers only going one way.

 Also, the node join validation can now selectively fail specific nodes (previously the entire batch was failed). This is an important preparation for a follow up PR where we plan to have a rejected joining node die with dignity.

### CODE REVIEW

**Comment by bleskes** [comment]
@clintongormley any tips on where to add docs for this?

**Comment by clintongormley** [comment]
@bleskes under rolling upgrades? https://www.elastic.co/guide/en/elasticsearch/reference/master/rolling-upgrades.html

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java [hunk 1]
```diff
@@ -668,7 +668,7 @@ public DiscoveryNodes build() {
             ImmutableOpenMap.Builder<String, DiscoveryNode> masterNodesBuilder = ImmutableOpenMap.builder();
             ImmutableOpenMap.Builder<String, DiscoveryNode> ingestNodesBuilder = ImmutableOpenMap.builder();
             Version minNodeVersion = Version.CURRENT;
-            Version maxNodeVersion = Version.CURRENT;
+            Version maxNodeVersion = null;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java]: I think I'm confused why there's an asymmetry between min and max, would you please explain (or fix)?

File: core/src/test/java/org/elasticsearch/VersionTests.java [hunk 2]
```diff
@@ -343,6 +347,20 @@ public void testIsCompatible() {
...
+
+        Version a = randomVersion(random());
+        Version b = randomVersion(random());
+        assertThat(a.isCompatible(b), equalTo(b.isCompatible(a)));
+    }
+
+    /* tests that if a new version's minCompatVersion is always equal or higher to any older version */
+    public void testMinCompatVersionOrderRespectsVersionOrder() {
+        List<Version> versionsByMinCompact = new ArrayList<>(allVersions());
+        versionsByMinCompact.sort(Comparator.comparing(Version::minimumCompatibilityVersion));
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/VersionTests.java]: Nit: `Compact` -> `Compat`

File: core/src/test/java/org/elasticsearch/discovery/zen/NodeJoinControllerTests.java [hunk 3]
```diff
@@ -585,6 +585,48 @@ public void testRejectingRestartedNodeJoinsBeforeProcessingNodeLeft() throws Int
...
+        } else {
+            goodJoin = joinNodeAsync(goodNode);
+            badJoin = joinNodeAsync(badNode);
+        }
+        assert goodJoin.isDone() == false;
+        assert badJoin.isDone() == false;
+        latch.countDown();
+        goodJoin.get();
+        ExecutionException e = expectThrows(ExecutionException.class, badJoin::get);
+        assertThat(e.getCause(), instanceOf(IllegalStateException.class));
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/discovery/zen/NodeJoinControllerTests.java]: Please assert on the message (to guard against `IllegalStateException` being thrown for another reason).

**Review by jasontedor — COMMENTED** [review]
It looks fine, I left a few comments.

**Review by s1monw — APPROVED** [review]
LGTM given @jasontedor comments are addressed. Thanks @bleskes

File: core/src/main/java/org/elasticsearch/discovery/zen/MembershipAction.java [hunk 4]
```diff
@@ -207,6 +209,31 @@ static void ensureIndexCompatibility(final Version nodeVersion, MetaData metaDat
...
     }
 
+    /** ensures that the joining node has a version that's compatible all nodes in the cluster */
+    static void ensureNodesCompatibility(final Version joiningNodeVersion, DiscoveryNodes currentNodes) {
+        final Version minNodeVersion = currentNodes.getMinNodeVersion();
+        final Version maxNodeVersion = currentNodes.getMaxNodeVersion();
+        ensureNodesCompatibility(joiningNodeVersion, minNodeVersion, maxNodeVersion);
+    }
+
+    /** ensures that the joining node has a version that's compatible all nodes in the cluster */
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/MembershipAction.java]: corrected. thanks

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java (hunk 1 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java]: yeah, that has to do with the fact that an empty builder (which can't really exist, but that's not clear from this code) will report min/max as CURRENT. This means that you can init the min to the highest possible value (current) but the max can't as CURRENT is higher than anything else you might encounter in the nodes. I'll just change the code to consistent and avoid confusion.

**Comment by bleskes** [comment]
@jasontedor thanks, care to take another look?

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java]: Thanks.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by bleskes** [comment]
Thx @jasontedor @s1monw & @clintongormley

**Comment by bleskes** [comment]
I reverted this change due to test failures. Will look at it tomorrow

**Comment by bleskes** [comment]
this is superseded by https://github.com/elastic/elasticsearch/pull/25808 , which also contains a fix to the failures.

**Pull request merged**